src/lib/optimizer/strategy/constant_calculation_rule.cpp

```diff
@@ -47,54 +62,107 @@ bool ConstantCalculationRule::apply_to(const std::shared_ptr<AbstractLQPNode>& n
+      const auto column_reference = boost::get<LQPColumnReference>(predicate_node->value());
+      if (column_reference_to_value_map.find(column_reference) != column_reference_to_value_map.end()) {
+        // We replace the LQPColumnReference with the actual result of the Expression it was referring to.
+        auto new_predicate_node =
+            std::make_shared<PredicateNode>(predicate_node->column_reference(), predicate_node->predicate_condition(),
+                                            column_reference_to_value_map.at(column_reference));
+        predicate_node->replace_with(new_predicate_node);
+        tree_changed = true;
+      }
+    }
+  } else if (node->type() == LQPNodeType::Projection) {
+    auto projection_node = std::static_pointer_cast<ProjectionNode>(node);
+    auto new_column_expressions = std::vector<std::shared_ptr<LQPExpression>>();
+    auto expression_replaced = false;
+
+    for (const auto expression : projection_node->column_expressions()) {
+      if (expression->type() == ExpressionType::Column &&
+          column_reference_to_value_map.find(expression->column_reference()) != column_reference_to_value_map.end()) {
+        new_column_expressions.push_back(
+            LQPExpression::create_literal(column_reference_to_value_map.at(expression->column_reference())));
+        expression_replaced = true;
+      } else {
+        new_column_expressions.push_back(expression);
+      }
     }
 
-    parent_tree_changed |= _replace_expression_in_parents(parent, expression_column, value);
+    if (expression_replaced) {
+      auto new_projection_node = std::make_shared<ProjectionNode>(new_column_expressions);
+      projection_node->replace_with(new_projection_node);
+      tree_changed = true;
+    }
   }
-  return parent_tree_changed;
+
+  return tree_changed;
 }
 
-void ConstantCalculationRule::_remove_column_from_projection(const std::shared_ptr<ProjectionNode>& node,
-                                                             ColumnID column_id) {
-  auto column_expressions = node->column_expressions();
-  column_expressions.erase(column_expressions.begin() + column_id);
+bool ConstantCalculationRule::_remove_columns_from_projections(
+    const std::shared_ptr<AbstractLQPNode>& node,
+    const std::map<LQPColumnReference, AllTypeVariant>& column_reference_to_value_map) {
+  auto tree_changed = false;
+
+  if (node->left_child()) {
+    tree_changed |= _remove_columns_from_projections(node->left_child(), column_reference_to_value_map);
+  }
+  if (node->right_child()) {
+    tree_changed |= _remove_columns_from_projections(node->right_child(), column_reference_to_value_map);
+  }
+
+  if (node->type() == LQPNodeType::Projection) {
+    for (const auto& parent_node : node->parents()) {
+      // We only want to remove columns from Projections from which we know they are not referenced anymore
+      if (parent_node->type() != LQPNodeType::Predicate && parent_node->type() != LQPNodeType::Projection) {
```

Comment:
Why would we only remove projection nodes below Predicates and Projections? I think this needs explanation, in my mind this distinction doesn't make intuitive sense.

Reply:
Initially it was there to "narrow" this rule down a little bit, basically to cover only cases we know are safe. But I changed this anyway in the process of working on this, to only check if the output node is the Root node, because in that case the ProjectionNode is the actual result node, and there we don't want to remove these columns.